### PR TITLE
[Issue-480] Dependency Notations for Kermit

### PR DESCRIPTION
**Pull request by brady-aiello**
## 🚀 Description
The KMP logging library, Kermit, has reached 1.0, and I'm adding the Kermit logging dependencies.

## 📄 Motivation and Context

Kermit is the go-to solution for KMP logging. It's performant, configurable, easy to use, has bugsnag and crashlytics extensions, and an optional compiler plugin for removing debug log calls altogether in production. More info in the ticket: https://github.com/jmfayard/refreshVersions/issues/480

## 🧪 How Has This Been Tested?

- `./gradlew :refreshVersions:check`
- inspections that the coordinates are generated in `dependencies-mapping-validated.txt` and `bundled-dependencies-validated.txt`.
- .`/gradlew :refreshVersions:publishToMavenLocal` followed by testing compilation in an Android project using `refreshVersions`.

## 📦 Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have read the [guidelines for the development process](https://jmfayard.github.io/refreshVersions/contributing/submitting-prs/dev-process/)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Sample usage:
Screen Shot 2022-01-01 at 11 49 03 AM

### CODE REVIEW

File: plugins/dependencies/src/main/resources/refreshVersions-rules/dependency-groups-alias-rules.txt [hunk 1]
```diff
@@ -10,6 +10,9 @@ io.github.javaeden.orchid:*
 org.http4k:http4k-*
     ^^^^^^
 
+co.touchlab:kermit*
+            ^^^^^^^
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/refreshVersions-rules/dependency-groups-alias-rules.txt:14]:
```suggestion
co.touchlab:kermit(-*)
            ^^^^^^^^^^
```

File: plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
- * - [GitHub Releases here](https://github.com/touchlab/Stately/releases)
- */
 @Incubating
 object Touchlab {
+
+    /**
+     * Multithreaded Kotlin Multiplatform Utilities
+     *
+     * - GitHub repo: https://github.com/touchlab/Stately
+     * - Releases: https://github.com/touchlab/Stately/releases
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt:13]: We are using markdown links so they are clickable right from the IDE that renders the KDoc.

```suggestion
     * [GitHub releases](https://github.com/touchlab/Stately/releases)
     *
     * GitHub repo: [touchlab/Stately](https://github.com/touchlab/Stately)
```

File: plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt [hunk 3]
```diff
@@ -26,5 +26,32 @@ object Touchlab {
...
+     *
+     * Github repo: https://github.com/touchlab/Kermit
+     * Releases: https://github.com/touchlab/Kermit/releases
+     * 1.0 release blog: https://medium.com/@kpgalligan/kermit-kmp-logging-1-0-196bf799b738
+     */
+    object Kermit : DependencyGroup(
+        group = "co.touchlab",
+        rawRule = """
+        co.touchlab:kermit*
+                    ^^^^^^^
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt:45]:
```suggestion
        co.touchlab:kermit(-*)
                    ^^^^^^^^^^
```

File: plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt [hunk 4]
```diff
@@ -26,5 +26,32 @@ object Touchlab {
...
+    val kermit = Kermit
+
+    /**
+     * Kermit is a Kotlin Multiplatform logging utility with composable log outputs.
+     * The library provides prebuilt loggers for outputting to platform logging tools
+     * such as Logcat and NSLog.
+     *
+     * Github repo: https://github.com/touchlab/Kermit
+     * Releases: https://github.com/touchlab/Kermit/releases
+     * 1.0 release blog: https://medium.com/@kpgalligan/kermit-kmp-logging-1-0-196bf799b738
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt:39]:
```suggestion
     * [GitHub Releases](https://github.com/touchlab/Kermit/releases)
     *
     * Github repo: [touchlab/Kermit](https://github.com/touchlab/Kermit)
     *
     * [1.0 release blog](https://medium.com/@kpgalligan/kermit-kmp-logging-1-0-196bf799b738)
```

File: plugins/dependencies/src/main/resources/removed-dependencies-versions-keys.txt [hunk 5]
```diff
@@ -51,3 +51,16 @@ androidx.legacy..legacy-support-v4=version.androidx.legacy
...
+co.touchlab..kermit-crashlytics=version.kermit
+co.touchlab..kermit-gradle-plugin=version.kermit
+co.touchlab..kermit-test=version.kermit
+co.touchlab..kermit-bugsnag-test=version.kermit
+co.touchlab..kermit-bugsnag=version.kermit
+co.touchlab..kermit-crashlytics-test=version.kermit
+co.touchlab..kermit-crashlytics=version.kermit
+co.touchlab..kermit-gradle-plugin=version.kermit
+co.touchlab..kermit-test=version.kermit
+co.touchlab..kermit=version.kermit
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/removed-dependencies-versions-keys.txt:66]: Why are these in the removed file?

File: plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt [hunk 6]
```diff
@@ -300,6 +300,13 @@ app.cash.copper..copper-flow=version.app.cash.copper
 app.cash.copper..copper-rx2=version.app.cash.copper
 app.cash.copper..copper-rx3=version.app.cash.copper
 app.cash.turbine..turbine=version.app.cash.turbine
+co.touchlab..kermit-bugsnag-test=version.kermit-bugsnag-test
+co.touchlab..kermit-bugsnag=version.kermit-bugsnag
+co.touchlab..kermit-crashlytics-test=version.kermit-crashlytics-test
+co.touchlab..kermit-crashlytics=version.kermit-crashlytics
+co.touchlab..kermit-gradle-plugin=version.kermit-gradle-plugin
+co.touchlab..kermit-test=version.kermit-test
+co.touchlab..kermit=version.NO-RULE
```
Comment (LouisCAD) [inline plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt:309]: This one doesn't look right. The ones above also don't look right, unless you really have an independent version per artifact.

**Review by LouisCAD — COMMENTED** [review]
Thanks for the contribution!

I think I'll take supersede this PR with a commit on my machine so I can address the comments I added on several changed lines. I'll make sure you are the author in the git history.

File: plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt [hunk 7]
```diff
@@ -26,5 +27,34 @@ object Touchlab {
...
+     *
+     * Github repo: [touchlab/Kermit](https://github.com/touchlab/Kermit)
+     *
+     * [1.0 release blog](https://medium.com/@kpgalligan/kermit-kmp-logging-1-0-196bf799b738)
+     */
+    object Kermit : DependencyGroup(
+        group = "co.touchlab",
+        rawRule = """
+        co.touchlab:kermit(-*)
+                    ^^^^^^^^^^
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt:48]:
```suggestion
        co.touchlab:kermit(-*)
                    ^^^^^^
```

File: plugins/dependencies/src/main/resources/refreshVersions-rules/dependency-groups-alias-rules.txt [hunk 8]
```diff
@@ -10,6 +10,9 @@ io.github.javaeden.orchid:*
 org.http4k:http4k-*
     ^^^^^^
 
+co.touchlab:kermit(-*)
+            ^^^^^^^^^^
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/resources/refreshVersions-rules/dependency-groups-alias-rules.txt:14]:
```suggestion
co.touchlab:kermit(-*)
            ^^^^^^
```

File: plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt (hunk 6 above)
Comment (LouisCAD) [reply plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt:309]: Found out the culprit. The code that reads the `rawRule` of `DependencyGroup`s before putting them in the `dependency-groups-alias-rules.txt` file was a regular tests, which meant it didn't run in time. I made it run `@BeforeAll` instead, so it won't happen again, I'll push in a few minutes.

**Comment by brady-aiello** [comment]
> Thanks for the contribution!
> 
> I think I'll take supersede this PR with a commit on my machine so I can address the comments I added on several changed lines. I'll make sure you are the author in the git history.

Sounds good! Thanks!

File: plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt (hunk 6 above)
Comment (LouisCAD) [reply plugins/dependencies/src/test/resources/dependencies-versions-key-validated.txt:309]: Took me more than a few minutes to push the entire thing (I had WIP dependency notation removals to finish dealing with), but it's there at last!

**Comment by LouisCAD** [comment]
Thank you!

**Comment by brady-aiello** [comment]
Yew! 🤙

File: plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt [hunk 9]
```diff
@@ -24,5 +27,34 @@ object Touchlab : IsNotADependency {
...
+     * [1.0 release blog](https://medium.com/@kpgalligan/kermit-kmp-logging-1-0-196bf799b738)
+     */
+    object Kermit : DependencyGroup(
+        group = "co.touchlab",
+        rawRule = """
+        co.touchlab:kermit(-*)
+                    ^^^^^^
+        """.trimIndent()
+    ) {
+        val kermit = module("kermit")
```
Comment (LouisCAD) [inline plugins/dependencies/src/main/kotlin/dependencies/Touchlab.kt:51]: FYI, you could have replaced that `val kermit` that leads to `Touchlab.kermit.kermit` at use-site with making the `Kermit` object extend `DependencyNotationAndGroup`, providing the `name`, which allows to write `Touchlab.kermit` at use-site. I made the change on main.

**Pull request closed without merging**